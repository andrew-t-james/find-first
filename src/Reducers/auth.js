export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case 'GOOGLE_LOGIN':
      return {
        id: action.uid
      };
    case 'GOOGLE_LOGOUT':
      return {};
    default:
      return state;
  }
};