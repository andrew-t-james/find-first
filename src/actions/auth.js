export const googleSignInAction = user => ({
  type: 'GOOGLE_LOGIN',
  user
});

export const googleSignOutAction = () => ({
  type: 'GOOGLE_LOGOUT'
});