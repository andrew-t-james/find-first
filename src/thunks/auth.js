import { firebase, googleAuthProvider } from '../firebase/firebase';
import { googleSignInAction, googleSignOutAction } from '../Actions/auth';

export const googleLogin = () => {
  return async dispatch => {
    const googleUser = await firebase.auth().signInWithPopup(googleAuthProvider);
    dispatch(googleSignInAction(googleUser.user));
  };
};

export const googleLogout = () => {
  return async dispatch => {
    await firebase.auth().signOut();
    dispatch(googleSignOutAction());
  };
};