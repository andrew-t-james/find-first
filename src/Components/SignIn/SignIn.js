import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { googleLogin, googleLogout } from '../../thunks/auth';

export class SignIn extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: ''
    };
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  handleGoogleLogin = () => {
    this.props.googleLogin();
  }

  handleGoogleSignOut = () => {
    this.props.googleLogout();
  }

  render() {
    const { email, password } = this.state;

    return (
      <section>
        <form action="">
          <input
            type="text"
            name="email"
            value={email}
            onChange={this.handleChange}
          />
          <input
            type="password"
            name="password"
            value={password}
            onChange={this.handleChange}
          />
        </form>
        <button
          className="google"
          onClick={this.handleGoogleLogin}
        >
          Google
        </button>
        <button onClick={this.handleGoogleSignOut}>Sign Out</button>
      </section>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  googleLogin: () => dispatch(googleLogin()),
  googleLogout: () => dispatch(googleLogout())
});

export default connect(null, mapDispatchToProps)(SignIn);

SignIn.propTypes = {
  googleLogin: PropTypes.func
};