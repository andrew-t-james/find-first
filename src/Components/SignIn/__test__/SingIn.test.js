import React from 'react';
import { shallow } from 'enzyme';
import { SignIn, mapDispatchToProps } from '../SignIn';
import { googleSignInAction } from '../../../Actions/auth';

jest.mock('../../../firebase/firebase.js');

describe('<SignIn />', () => {
  let wrapper;
  const mockGoogleLogin = jest.fn();
  beforeEach(() => wrapper = shallow(
    <SignIn
      googleLogin={mockGoogleLogin}
    />
  ));

  test('should update email state on change', () => {
    const mockEvent = {
      target: {
        name: 'email',
        value: 'me@mail.com'
      }
    };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('email')).toEqual('me@mail.com');
  });

  test('should update password state on change', () => {
    const mockEvent = {
      target: {
        name: 'password',
        value: 'weak'
      }
    };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('password')).toEqual('weak');
  });

  test('should call handleGoogleLogin onClick', () => {
    wrapper.find('.google').simulate('click');

    expect(mockGoogleLogin).toHaveBeenCalled();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('mapDispatchToPops', () => {
    test('should call dispatch googleLogin is called', () => {
      const mockDispatch = jest.fn();
      const id = 1;
      const name = 'Steve';
      const mockUser = {
        uid: id,
        displayName: name
      };
      const actionToDispatch = googleSignInAction(mockUser);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.googleLogin();

      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});
