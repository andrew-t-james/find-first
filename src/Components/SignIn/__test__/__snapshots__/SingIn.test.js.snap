// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> should match snapshot 1`] = `
<section>
  <form
    action=""
  >
    <input
      name="email"
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </form>
  <button
    className="google"
    onClick={[Function]}
  >
    Google
  </button>
  <button
    onClick={[Function]}
  >
    Sign Out
  </button>
</section>
`;
